doc/source/reference/c-api/array.rst

```diff
@@ -2884,11 +3172,279 @@ Data-type descriptors
+
+   Returns the default descriptor instance for the DType. Must be
+   defined for parametric data types. Non-parametric data types return
+   the singleton by default.
+
+.. c:macro:: NPY_DT_common_dtype
+
+.. c:type:: PyArray_DTypeMeta *(PyArrayDTypeMeta_CommonDType)( \
+                PyArray_DTypeMeta *dtype1, PyArray_DTypeMeta *dtype2)
+
+   Given two input DTypes, determines the appropriate "common" DType
+   that can store values for both types. Returns ``Py_NotImplemented``
+   if no such type exists.
+
+.. c:macro:: NPY_DT_common_instance
+
+.. c:type:: PyArray_Descr *(PyArrayDTypeMeta_CommonInstance)( \
+               PyArray_Descr *dtype1, PyArray_Descr *dtype2)
+
+   Given two input descriptors, determines the appropriate "common"
+   descriptor that can store values for both instances. Returns ``NULL``
+   on error.
+
+.. c:macro:: NPY_DT_ensure_canonical
+
+.. c:type:: PyArray_Descr *(PyArrayDTypeMeta_EnsureCanonical)( \
+                PyArray_Descr *dtype)
+
+   Returns the "canonical" representation for a descriptor instance. The
+   notion of a canonical descriptor generalizes the concept of byte
+   order, in that a canonical descriptor always has native byte
+   order. If the descriptor is already canonical, this function returns
+   a new reference to the input descriptor.
+
+.. c:macro:: NPY_DT_setitem
+
+.. c:type:: int(PyArrayDTypeMeta_SetItem)(PyArray_Descr *, PyObject *, char *)
+
+   Implements scalar setitem for an array element given a PyObject.
+
+.. c:macro:: NPY_DT_getitem
+
+.. c:type:: PyObject *(PyArrayDTypeMeta_GetItem)(PyArray_Descr *, char *)
+
+   Implements scalar getitem for an array element. Must return a python
+   scalar.
+
+.. c:macro:: NPY_DT_get_clear_loop
+
+   If defined, sets a traversal loop that clears data in the array. This
+   is most useful for arrays of references that must clean up array
+   entries before the array is garbage collected. Implements
+   ``PyArrayMethod_GetLoop``.
+
+.. c:macro:: NPY_DT_get_fill_zero_loop
+
+   If defined, sets a traversal loop that fills an array with "zero"
+   values, which may have a DType-specific meaning. This is called
+   inside `numpy.zeros` for arrays that need to write a custom sentinel
+   value that represents zero if for some reason a zero-filled array is
+   not sufficient. Implements ``PyArrayMethod_GetLoop``.
+
+.. c:macro:: NPY_DT_finalize_descr
+
+.. c:type:: PyArray_Descr *(PyArrayDTypeMeta_FinalizeDescriptor)( \
+                PyArray_Descr *dtype)
+
+   If defined, a function that is called to "finalize" a descriptor
+   instance after an array is created. One use of this function is to
+   force newly created arrays to have a newly created descriptor
+   instance, no matter what input descriptor is provided by a user.
+
+PyArray_ArrFuncs slots
+^^^^^^^^^^^^^^^^^^^^^^
+
+In addition the above slots, the following slots are exposed to allow
+filling the ``PyArray_ArrFuncs`` struct attached to descriptor
```

Comment:
Would be good to have a working link to `PyArray_ArrFuncs` here. Also, are these still necessary, or just there for backwards compatibility? (e.g., `dotfunc` seems pretty pointless, but `compare` likely useful)

Reply:
Added a note that this will likely change in the future.

Reply:
ah, missed that - so ignore my request just now to add such a note!